Feature : Collateralization oracle

eswak opened the pull request:
Hi,

Here is a pull request to add a Collateralization Oracle as discussed [on the governance forums](https://tribe.fei.money/t/ethreweightqueuer-proposal-to-improve-reweight-mechanism-maintain-peg/1521/41).

This contract is pretty straightforward & mainly composed of pure functions : 
- It interacts with `EthUniswapPCVDeposit` to get the `totalValue()` and `getReserves()`.
- Circulating FEI supply is defined as FEI `totalSupply()` minus FEI in the Uniswap pool owned by `EthUniswapPCVDeposit`.
- `UniswapOracle` is used for the ETH/USD current price.
- `UniswapOracle` is used for the ETH/FEI current price (234b68f16b803c0dba7785565173dc8ca40c8159).
- It should be upgraded as PCV is deployed elsewhere in the future (as in #95).

I also implemented a small test suite inspired from the one of `UniswapOracle` :
Can be run with ` npm run test -- -g CollateralizationOracle` 
![image](https://user-images.githubusercontent.com/6916865/114373571-044b5900-9b83-11eb-9fa5-e6a0019e1990.png)

Also tested on a local mainnet fork, with results consistent with what's observed on https://app.fei.money/analytics (see comments below)

Best,

Joeysantoro commented on contracts/oracle/CollateralizationOracle.sol:
```diff
@@ -0,0 +1,109 @@
+    /// @notice get the current circulating suply of FEI
+    /// @return number of FEI in circulation
+    function _circulatingFei() internal view returns(uint256) {
+      (uint256 feiInPool, uint256 ethInPool) = ethUniswapPCVDeposit.getReserves();
+      uint256 ethPcv = ethUniswapPCVDeposit.totalValue();
+      uint256 feiPcv = Decimal.ratio(ethPcv, ethInPool).mul(feiInPool).asUint256();
```
Using the spot reserves of Uniswap would be vulnerable to flash manipulation. I'd suggest using an ETH-FEI TWAP oracle and then back out manipulation resistant numbers for the reserves by doing `k = feiInPool * ethInPool`
`resistantEthInPool = sqrt(k/ethFeiOraclePrice)` and `resistantFeiInPool = k/resistantEthInPool`

Here is the address for our current ETH-FEI TWAP: 0xB58cbCb1e98c7F3771847cc2D730bF9589b534e4. Its 30 min so would only be more useful for less price sensitive applications

eswak replied:
done

Joeysantoro reviewed (changes requested):
This is awesome thanks for coding it up so fast!

One final note is it is good to reduce the number of external contract calls as much as possible to save on gas. One call to read should only hit every external call once in the entire flow.

eswak commented on contracts/oracle/CollateralizationOracle.sol:
```diff
@@ -0,0 +1,109 @@
+import "./UniswapOracle.sol";
+import "../refs/CoreRef.sol";
+import "../external/SafeMathCopy.sol";
+import "../pcv/EthUniswapPCVDeposit.sol";
```
I used `UniswapPCVDeposit` that is also a `IUniRef` to be able to call `getReserves()`.

eswak commented on contracts/oracle/CollateralizationOracle.sol:
```diff
@@ -0,0 +1,109 @@
+pragma solidity ^0.6.0;
+pragma experimental ABIEncoderV2;
+
+import "./UniswapOracle.sol";
```
done

eswak commented on contracts/oracle/CollateralizationOracle.sol:
```diff
@@ -0,0 +1,109 @@
+      uint256 ethPcv = ethUniswapPCVDeposit.totalValue();
+      uint256 feiPcv = Decimal.ratio(ethPcv, ethInPool).mul(feiInPool).asUint256();
+
+      return Decimal.from(fei().totalSupply()).sub(feiPcv).asUint256();
```
done

eswak commented:
I changed the implementation of `_collateralizationRatio()` to make the external calls only once.

But since we introduced the `ethFeiPriceOracle`, there are still 5 external calls to get the collateral ratio. Should we have an internal variable to keep the last value, and rely more on update() calls ?

On a related topic : is the code of the webapp (https://app.fei.money/analytics) available somewhere ? The analytics page could be updated to read the oracle.

Joeysantoro commented:
> On a related topic : is the code of the webapp (https://app.fei.money/analytics) available somewhere ? The analytics page could be updated to read the oracle.

The webapp code is not available but we could update it! Have you tested this oracle on a mainnet fork?

eswak commented:
Note : to be updated after #95 is merged (`EthReserveStabilizer` and `EthPCVDripper` will have non-zero balance of ETH that must be counted as part of the PCV).

eswak commented on contracts/oracle/CollateralizationOracle.sol:
```diff
@@ -0,0 +1,164 @@
+    /// @return true if the protocol is currently overcollateralized
+    function overCollateralized() external view returns(bool) {
+      (Decimal.D256 memory ratio,) = _collateralizationRatio();
+      return ratio.mul(1000000000000000000).asUint256() > 1000000000000000000;
```
done

eswak commented on contracts/oracle/CollateralizationOracle.sol:
```diff
@@ -0,0 +1,164 @@
+    /// @return collateralization ratio
+    /// @dev price is to be denominated in percentage, with 18 decimals (1.25e18 for 125%)
+    function collateralizationRatio() external view returns(uint256) {
+      (Decimal.D256 memory ratio,) = _collateralizationRatio();
```
done

eswak commented on contracts/oracle/CollateralizationOracle.sol:
```diff
@@ -0,0 +1,164 @@
+        bool updated = ethUsdPriceOracle.update();
+        updated = updated || ethFeiPriceOracle.update();
+
+        (Decimal.D256 memory ratio,) = _collateralizationRatio();
+
+        emit Update(ratio.asUint256());
```
done (the `UniswapOracle` always return true, that's why I didn't bother, but their implementation may change)
